[DOC-252] Added pg_locks and moved some pages from query tuning to observability (#21231)
yugabyte/yugabyte-db: docs/content/stable/explore/observability/pg-locks.md

---

> title: Get lock information insights with pg_locks

Reviewer: need to think of a better title

linkTitle: Get lock information insights
description: Using pg_locks to get insights into lock information specific to YugabyteDB's distributed SQL architecture.
headerTitle: Get lock information insights with pg_locks
menu:
  stable:
    identifier: pg-locks
    parent: explore-observability
    weight: 410
type: docs
---

YugabyteDB supports the PostgreSQL [pg_locks](https://www.postgresql.org/docs/current/view-pg-locks.html) system view that provides information about the locks held and requested by the current active transactions in a database. YugabyteDB enhances this view with two new fields `waitend` and `ybdetails` offering insights into lock information specific to YugabyteDB's distributed architecture. The `pg_locks` view is tailored to YugabyteDB's lock handling mechanisms, providing a comprehensive overview of database lock states.

> 

Author: examples are at the end of the page.

| waitstart | timestampz | Time at which a YB-TServer starts waiting for this lock. |
| waitend | timestampz | Time at which a lock gets acquired. |
| ybdetails | JSONB | Field with details specific to YugabyteDB locks, including `node`, `transactionid`, and `blocked_by details`.|

Note that certain Postgres-specific fields such as `page`, `tuple`, `virtualxid`, `transactionid`, `classid`, `objid`, `objsubid`, `virtualtransaction` are not applicable in YugabyteDB and are always NULL.

## YugabyteDB-specific changes

The `pg_locks` view includes the following new YugabyteDB-specific fields:

### waitend

The waitend field contains the timestamp at which the lock was acquired by a transaction (PostgtreSQL does not record this timestamp currently.)

### ybdetails

The ybdetails field is a JSONB type that encapsulates additional information about each lock, specific to YugabyteDB's distributed architecture, and includes the following attributes:

> - `node`: The YB-TServer UUID of the node hosting the PostgreSQL backend that is holding the lock. It helps in identifying the specific node in the YugabyteDB cluster where the lock is being held, facilitating easier troubleshooting and monitoring of lock distribution across nodes.

Reviewer: ```suggestion
- `node`: The YB-TServer UUID of the node hosting the PostgreSQL backend that is holding the lock. Identifies the specific node in the YugabyteDB cluster where the lock is being held, which allows you to troubleshoot and monitor lock distribution across nodes.
```

> - `transactionid`: The UUID of the YugabyteDB transaction ID owning this lock. This field is NULL for single shard or fastpath operations, offering visibility into the transactional context of the lock. It aids in tracking and managing transactions across the distributed database.

Reviewer: ```suggestion
- `transactionid`: The UUID of the YugabyteDB transaction ID owning this lock. This field is NULL for single shard or fastpath operations, offering visibility into the transactional context of the lock. Use this field to track and manage transactions across the distributed database.
```

- `subtransaction_id`: The ID of the subtransaction in which the lock was acquired.
- `is_explicit`: True when the lock was acquired explicitly, such as through FOR UPDATE, FOR NO KEY UPDATE, FOR SHARE, FOR KEY SHARE, and so on. This attribute helps distinguish between locks acquired automatically by the database system and those requested explicitly by you, aiding in lock analysis and optimization efforts.
- `tablet_id`: The ID of the tablet containing this lock. In YugabyteDB, data is sharded into tablets, and this ID helps identify the specific shard where the lock exists, crucial for diagnosing sharding-related lock contention issues.
- `blocked_by`: A list of transactions blocking the acquisition of this lock. This field is helpful in identifying deadlock scenarios, and determining which transaction is blocking other operations from moving forward.

> - `keyrangedetails`: Provides details about what keys the lock is held, including the following:

Reviewer: "what keys the lock is held" this fragment doesn't quite parse

>   - `cols`: A list of column values from the PRIMARY KEY of the table, offering insight into the exact row(s), or key range locked.

Reviewer: ```suggestion
  - `cols`: A list of column values from the PRIMARY KEY of the table, offering insight into the exact row(s) or key range that is locked.
```

  - `attnum`: The PostgreSQL attribute number indicating if the lock is a column-level lock, linking the lock to the specific table column.
  - `column_id`: The column ID in DocDB if the lock is a column-level lock, further aligning lock information with YugabyteDB's internal document-oriented architecture.
  - `multiple_rows_locked`: Indicates when the lock is held on more than one entry in DocDB, helping to understand the scope of the lock in the database's document model.

## Configurable parameters for lock management

> YugabyteDB offers several GUC parameters to customize how locks are queried and displayed, so you can tailor the lock information as needed. `yb_locks_min_txn_age` and `yb_locks_max_transactions`, control the filtering and limitation of transactions in lock queries.

Reviewer: ```suggestion
YugabyteDB offers several GUC parameters to customize how locks are queried and displayed, so you can tailor the lock information as needed. `yb_locks_min_txn_age` and `yb_locks_max_transactions` control the filtering and limitation of transactions in lock queries.
```

### yb_locks_min_txn_age

> The `yb_locks_min_txn_age` parameter specifies the minimum age of a transaction (in seconds) before its locks gets included in the results returned from querying the `pg_locks` view. By setting this parameter, you can focus on older transactions that may be more relevant to performance tuning or deadlock resolution efforts. Transactions that are started more recently than the specified duration is not shown, helping to reduce clutter and focus on potentially problematic transactions.

Reviewer: ```suggestion
The `yb_locks_min_txn_age` parameter specifies the minimum age of a transaction (in seconds) before its locks are included in the results returned from querying the `pg_locks` view. By setting this parameter, you can focus on older transactions that may be more relevant to performance tuning or deadlock resolution efforts. Transactions that are started more recently than the specified duration are not shown, helping to reduce clutter and focus on potentially problematic transactions.
```

Default: 1 second

> Example: You can use the following command to adjust the minimum transaction age to 5 seconds, thereby excluding more recent transactions from lock queries:

Reviewer: ```suggestion
Example: Enter the following command to change the minimum transaction age to 5 seconds:
```

```sh
SET session yb_locks_min_txn_age = 5000;
```

### yb_locks_max_transactions

> The `yb_locks_max_transactions` parameter sets the maximum number of transactions for which lock information is displayed when you query the pg_locks view. It allows to limit the output to the most relevant transactions, which is particularly beneficial in environments with high levels of concurrency and transactional activity. By controlling the volume of information returned, this parameter helps in managing the analysis of lock contention more effectively.

Reviewer: ```suggestion
The `yb_locks_max_transactions` parameter sets the maximum number of transactions for which lock information is displayed when you query the pg_locks view. You can limit the output to the most relevant transactions, which is particularly beneficial in environments with high levels of concurrency and transactional activity. By controlling the volume of information returned, this parameter helps in managing the analysis of lock contention more effectively.
```

Default: 16

> Example: You can use the following command to change the maximum number of transactions to display to 10:

Reviewer: ```suggestion
Example: Enter the following command to change the maximum number of transactions to display to 10:
```

```sh
SET session yb_locks_max_transactions = 10;
```

## Examples

> Following are some examples of how to use the pg_locks view in YugabyteDB.

Reviewer: ```suggestion
The following examples show how you can use the pg_locks view in YugabyteDB:
```

> {{% explore-setup-single %}}

Reviewer: move this above the paragraph
